CONTRIBUTING.md

```diff
@@ -456,6 +456,188 @@ checks if contributed code does not have the appropriate license headers.
+the values in `[...]` to help distinguish them from the static part of the
+message:
+
+    logger.debug("operation failed [{}] times in [{}]ms", failureCount, elapsedMillis);
+
+You can also pass in an exception to log it including its stack trace (and
+causes and so on...):
+
+    logger.debug("operation failed", exception);
+
+If you wish to use placeholders and an exception at the same time, construct a
+`ParameterizedMessage`:
+
+    logger.debug(new ParameterizedMessage("failed at offset [{}]", offset), exception);
+
+You can also use a `Supplier<ParameterizedMessage>` to avoid constructing
+expensive messages that will usually be discarded:
+
+    logger.debug(() -> new ParameterizedMessage("rarely seen output [{}]", expensiveMethod()));
+
+Logging is an important behaviour of the system and sometimes deserves its own
+unit tests, especially if there is complex logic for computing what is logged
+and when to log it. You can use a `org.elasticsearch.test.MockLogAppender` to
+make assertions about the logs that are being emitted.
+
+Logging is a powerful diagnostic technique but it is not the only possibility.
+You should also consider exposing information about your component via an API
+instead of in logs. For instance you can implement APIs to report its current
+status, various statistics, and maybe even details of recent failures.
+
+#### Log levels
+
+Each log message is written at a particular _level_. By default Elasticsearch
+will suppress messages at the two most verbose levels, `TRACE` and `DEBUG`, and
+will output messages at all other levels. Users can configure which levels of
+message are written by each logger at runtime, but you should expect everyone
+to run with the default configuration almost all of the time and choose your
+levels accordingly.
+
+##### `TRACE`
+
+This is the most verbose level, disabled by default, and it is acceptable if it
+generates a very high volume of logs. The target audience of `TRACE` logs
+comprises developers who are trying to deeply understand some unusual runtime
+behaviour of a system. For instance `TRACE` logs may be useful when
+understanding an unexpected interleaving of concurrent actions or some
+unexpected consequences of a delayed response from a remote node.
+
+`TRACE` logs will normally only make sense when read alongside the code, and
+typically they will be read as a whole sequence of messages rather than in
+isolation.
+
+Even though `TRACE` logs may be very verbose, you should still exercise some
+judgement when deciding when to use them. In many cases it will be easier to
+understand the behaviour of the system using tests or by analysing the code
+itself rather than by trawling through hundreds of trivial log messages.
+
+It may not be easy, or even possible, to obtain `TRACE` logs from a production
+system. Therefore they are not appropriate for information that you would
+expect to be useful in diagnosing problems in production.
```

Comment:
I'm kind of unsure why you'd commit logs for some reason *other* than to debug stuff in production. I guess to debug test failures? I feel like TRACE logs are useful to debug some issues in production. But super rarely. At least, that makes sense given all the other things you mentioned - them being difficult to access in production and very very chatty.

Reply:
> I guess to debug test failures?

Yeah either actual automated tests or cases where you can reproduce the problem in a non-production cluster but need the extra detail to see what's really happening. I use them particularly where whatever I'm chasing is super-rare, so I have to leave a test running for hours until it finally happens.

Maybe look at this the other way round: if (volume-wise) it's safe to log certain things in production, it can probably go to level `DEBUG`. If the volume is too high then it's going to be risky enabling them on a production cluster both in terms of the effects on the cluster itself and on whoever is consuming the logs. In any case we're not saying you can't use `TRACE` logs in prod, just that it might not be easy or even possible. Some customers may have policies that completely forbid use of the `TRACE` logging level in their production environments, for instance.

Reply:
I think the phrase

> you would expect to be useful

is key here, and somewhat subjective.  
It can range from "If I try hard enough, I could come up with some exceptional scenario under which this information might help me with diagnosis" to "when this very common problem happens, I need to know this information in order to diagnose it".

For the latter, the ideal would be to detect the problematic scenario directly, and log at `WARN`.

But I wouldn't assume that "I can come up with a situation under which this might be useful" is the measure we're looking for. Because that's almost always going to be true.

I have no suggestions, and I don't even think there's anything wrong with the proposed text, I just think the way that different people read that phrase will affect whether they see it as too strict.

Reply:
++ I did `s/expect to be useful/normally expect to be useful/` in 6942b041d2a13741434689ca599f49e811ceb858 and also added a note to discuss these things with reviewers when in doubt.